zcash_primitives/src/transaction/components.rs

```diff
@@ -0,0 +1,302 @@
+    redjubjub::{PublicKey, Signature},
+};
+use std::io::{self, Read, Write};
+
+use serialize::Vector;
+use JUBJUB;
+
+// π_A + π_B + π_C
+const GROTH_PROOF_SIZE: usize = (48 + 96 + 48);
+// π_A + π_A' + π_B + π_B' + π_C + π_C' + π_K + π_H
+const PHGR_PROOF_SIZE: usize = (33 + 33 + 65 + 33 + 33 + 33 + 33 + 33);
+
+const ZC_NUM_JS_INPUTS: usize = 2;
+const ZC_NUM_JS_OUTPUTS: usize = 2;
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct Amount(pub i64);
+
+pub struct Script(pub Vec<u8>);
+
+impl Script {
+    pub fn read<R: Read>(mut reader: R) -> io::Result<Self> {
+        let script = Vector::read(&mut reader, |r| r.read_u8())?;
+        Ok(Script(script))
+    }
+
+    pub fn write<W: Write>(&self, mut writer: W) -> io::Result<()> {
+        Vector::write(&mut writer, &self.0, |w, e| w.write_u8(*e))
+    }
+}
+
+pub struct OutPoint {
+    hash: [u8; 32],
+    n: u32,
+}
+
+impl OutPoint {
+    pub fn read<R: Read>(mut reader: R) -> io::Result<Self> {
+        let mut hash = [0; 32];
+        reader.read_exact(&mut hash)?;
+        let n = reader.read_u32::<LittleEndian>()?;
+        Ok(OutPoint { hash, n })
+    }
+
+    pub fn write<W: Write>(&self, mut writer: W) -> io::Result<()> {
+        writer.write_all(&self.hash)?;
+        writer.write_u32::<LittleEndian>(self.n)
+    }
+}
+
+pub struct TxIn {
+    pub prevout: OutPoint,
+    script_sig: Script,
+    pub sequence: u32,
+}
+
+impl TxIn {
+    pub fn read<R: Read>(mut reader: &mut R) -> io::Result<Self> {
+        let prevout = OutPoint::read(&mut reader)?;
+        let script_sig = Script::read(&mut reader)?;
+        let sequence = reader.read_u32::<LittleEndian>()?;
+
+        Ok(TxIn {
+            prevout,
+            script_sig,
+            sequence,
+        })
+    }
+
+    pub fn write<W: Write>(&self, mut writer: W) -> io::Result<()> {
+        self.prevout.write(&mut writer)?;
+        self.script_sig.write(&mut writer)?;
+        writer.write_u32::<LittleEndian>(self.sequence)
+    }
+}
+
+pub struct TxOut {
+    value: Amount,
+    script_pubkey: Script,
+}
+
+impl TxOut {
+    pub fn read<R: Read>(mut reader: &mut R) -> io::Result<Self> {
+        let value = Amount(reader.read_i64::<LittleEndian>()?);
+        let script_pubkey = Script::read(&mut reader)?;
+
+        Ok(TxOut {
+            value,
+            script_pubkey,
+        })
+    }
+
+    pub fn write<W: Write>(&self, mut writer: W) -> io::Result<()> {
+        writer.write_i64::<LittleEndian>(self.value.0)?;
+        self.script_pubkey.write(&mut writer)
+    }
+}
+
+pub struct SpendDescription {
+    pub cv: edwards::Point<Bls12, Unknown>,
+    pub anchor: Fr,
+    pub nullifier: [u8; 32],
+    pub rk: PublicKey<Bls12>,
+    pub zkproof: [u8; GROTH_PROOF_SIZE],
+    pub spend_auth_sig: Signature,
+}
+
+impl SpendDescription {
+    pub fn read<R: Read>(mut reader: &mut R) -> io::Result<Self> {
+        let cv = edwards::Point::<Bls12, Unknown>::read(&mut reader, &JUBJUB)?;
+        let anchor = {
```

Comment:
Do we check this when deserializing `SpendDescription` in `zcashd`? I think we check it when we verify the transaction. Not sure if the distinction is relevant.

Reply:
And this applies to any other failing-checks. I think it's okay to check it all during deserialization but just want to point that out.

Reply:
You are correct that we don't check this during deserialisation in `zcashd` (we treat all fields there as binary blobs). However, we **do** enforce these checks when checking consensus rules (as this deserialisation happens in the `librustzcash` FFI wrappers), so there should not be any inconsistencies between these implementations in terms of what is allowed into the block chain.

Where I'm more concerned is around the prime-order subgroup checks `Point<_, Unknown>::as_prime_order()`. It's not immediately clear whether e.g. `epk` being `Unknown` here conflicts with pathways where it needs to be `PrimeOrder`. I'm currently assuming that the type system will ensure that implementing any consensus-critical code that requires `epk` to be `PrimeOrder` will be consistent via both the pure-Rust and `librustzcash` pathways, but I'd like some more review on that point.

Reply:
I don't have time to review the code right now, sorry. (Remind me next week.) The spec says that epk is of type KA<sup>Sapling</sup>.Public which is J (Jubjub point). In practice it will be ensured to be prime order when the output proof is checked, but it's not specified to be checked outside the circuit. I think that was intentional to avoid the cost of the check. It *is* specified to be a canonical encoding, like all serialized points.

Reply:
BTW, it's allowed from a consensus point of view to redundantly check something that is also checked by the proof, since that only changes which failure is reported. It's less efficient than relying on the proof, though.

Reply:
Also note that it is *valid*, according to the spec, for epk to be the zero point. This is allowed by the circuit when esk = 0 or r<sub>J</sub>.

Reply:
I continue to believe that naming the trait `PrimeOrder`, when it actually means "in the prime order subgroup", is *very* misleading.